include/envoy/thread_local/thread_local.h

```diff
@@ -46,6 +46,13 @@ class Slot {
    */
   virtual void runOnAllThreads(Event::PostCb cb) PURE;
 
+  /**
+   * Run a callback on all registered threads with a barrier.
+   * @param cb supplies the callback to run on each thread.
+   * @param main_callback supplies the callback to run on main thread after threads are done.
```

Comment:
Can you add a warning here that there is no guarantee that this callback will fire during shutdown. I think that's OK for now but I would like to clearly mark that in case we need that later.